backend/migrations/2023-07-03-165000_heatmap/up.sql

```diff
@@ -0,0 +1,241 @@
+                preference := 0.0;
+                relevance := 0.0;
+            END IF;
+
+            x := i;
+            y := j;
+
+            RETURN NEXT;
+        END LOOP;
+    END LOOP;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Scales to values between 0 and 1.
+--
+-- Preference input space: Any value.
+-- Relevance input space: >=0
+CREATE OR REPLACE FUNCTION scale_score(input SCORE)
+RETURNS SCORE AS $$
+DECLARE
+    score SCORE;
+BEGIN
+    score.preference := 1 / (1 + exp(-input.preference));       -- standard sigmoid, so that f(0)=0.5
+    score.relevance := (2 / (1 + exp(-input.relevance))) - 1;   -- modified sigmoid, so that f(0)=0
+    RETURN score;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate score for a certain position.
+--
+-- p_map_id       ... map id
+-- p_layer_ids[1] ... plant layer
+-- p_plant_id     ... id of the plant for which to consider relations
+-- date           ... date at which to generate the heatmap
+-- x_pos,y_pos    ... coordinates on the map where to calculate the score
+CREATE OR REPLACE FUNCTION calculate_score(
+    p_map_id INTEGER,
+    p_layer_ids INTEGER [],
+    p_plant_id INTEGER,
+    date DATE,
+    x_pos INTEGER,
+    y_pos INTEGER
+)
+RETURNS SCORE AS $$
+DECLARE
+    score SCORE;
+    plants SCORE;
+BEGIN
+    plants := calculate_score_from_relations(p_layer_ids[1], p_plant_id, date, x_pos, y_pos);
+
+    score.preference := plants.preference;
+    score.relevance := plants.relevance;
+
+    RETURN score;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate score using the plants relations and their distances.
+CREATE OR REPLACE FUNCTION calculate_score_from_relations(
+    p_layer_id INTEGER,
+    p_plant_id INTEGER,
+    date DATE,
+    x_pos INTEGER,
+    y_pos INTEGER
+)
+RETURNS SCORE AS $$
+DECLARE
+    plant_relation RECORD;
+    distance REAL;
+    weight REAL;
+    score SCORE;
+BEGIN
+    IF NOT EXISTS (SELECT 1 FROM layers WHERE id = p_layer_id AND type = 'plants') THEN
+        RAISE EXCEPTION 'Plant layer with id % not found', p_layer_id;
+    END IF;
+
+    score.preference := 0.0;
+    score.relevance := 0.0;
+
+    FOR plant_relation IN (SELECT * FROM get_plant_relations(p_layer_id, p_plant_id, date)) LOOP
+        -- calculate euclidean distance
+        distance := sqrt((plant_relation.x - x_pos)^2 + (plant_relation.y - y_pos)^2);
+
+        -- calculate weight based on distance
+        -- weight decreases between 1 and 0 based on distance
+        -- distance is squared so it decreases faster the further away
+        -- weight is halved at 50 cm away
+        weight := 1 / (1 + (distance / 50)^2);
+
+        -- update score based on relation
+        IF plant_relation.relation = 'companion' THEN
+            score.preference := score.preference + 0.5 * weight;
+            score.relevance := score.relevance + 0.5 * weight;
+        ELSE
+            score.preference := score.preference - 0.5 * weight;
+            score.relevance := score.relevance + 0.5 * weight;
+        END IF;
```

Comment:
As the 2nd calculation is the same in both branches, it could be moved outside of the if-else block to avoid code duplication and make it more obvious that the value of `relation` only affects `preference`, but not `relevance`.

```suggestion
        IF plant_relation.relation = 'companion' THEN
            score.preference := score.preference + 0.5 * weight;
        ELSE
            score.preference := score.preference - 0.5 * weight;
        END IF;
        score.relevance := score.relevance + 0.5 * weight;
```